staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -308,3 +346,23 @@ func (tw *baseTimeoutWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	}
 	return conn, rw, err
 }
+
+var _ io.ReadCloser = &timeoutReader{}
+
+// timeoutReader wraps a http.Request.Body (an io.ReadCloser) to make
+// read calls fail once the timeout has occurred.
+type timeoutReader struct {
+	io.ReadCloser
+	ctx context.Context
+}
+
+func (r *timeoutReader) Read(b []byte) (int, error) {
+	if err := r.ctx.Err(); err != nil {
+		return 0, err
+	}
```

Comment:
technically both goroutines `A` and `B` can find `ctx.Err` to be `nil` and then each will attempt a `Read` which will cause a data race. I think we should use a `timeoutReader` that is similar to `baseTimeoutWriter`

Reply:
Makes sense.